Update embedding and execution specs for memory64

bvisness opened the pull request:
Memory and table operations in the embedding section have been updated to use u64 instead of u32, reflecting the changes to `limits`. Various operations in the execution section were also updated to reflect `idxtype`'s inclusion in `memtype` and `tabletype`.

Additionally, I have switched the storage in memory and table instances to sequences instead of vectors. This is because vectors are restricted to a maximum length of 2^32, which is incompatible with i64 storage.

Finally, I have decided not to try and address #67 in this PR or the one to follow. I'll submit a separate PR to rename "index type" to "offset type" across the board after all other spec changes are done, and then we can decide if we actually want to do it or not.

See also #75, which updates the JS API.

bvisness committed:
Update memory/table validation/execution/embedding to u64

Memory and table operations in the embedding section have been updated
to use u64 instead of u32, reflecting the changes to `limits`. Various
operations in the execution section were also updated to reflect
`idxtype`'s inclusion in `memtype` and `tabletype`, as well as the
supporting validation rules.

rossberg commented on document/core/syntax/conventions.rst:
```diff
@@ -123,11 +123,9 @@ Vectors
 .. math::
    \begin{array}{lllll}
    \production{vector} & \vec(A) &::=&
-     A^n
-     & (\iff n < 2^{32})\\
```
Ah, removing this constraint is not correct, since vectors are used in many places throughout the spec that continue to be syntactically restricted to 32-bit length.

The correct fix is to change the content types of table and memory instances from vectors to unrestricted sequences `xyz*`.

bvisness replied:
Done. As far as I can tell, nothing else in the spec needs to change to support this, since constructs like `|foo|` and `foo[a:b]` still work with sequences.

rossberg commented on document/core/exec/modules.rst:
```diff
@@ -322,26 +322,26 @@ Growing :ref:`memories <syntax-meminst>`
+       \wedge & \X{len} \leq 2^{48} \\
+       \wedge & \idxtype~\limits = \meminst.\MITYPE \\
        \wedge & \limits' = \limits \with \LMIN = \X{len} \\
        \wedge & \vdashlimits \limits' \ok) \\
```
I just noticed that this is outdated, and has been for a while. Instead of checking the length separately above, this should reuse type memtype validation, such that this line reads:
```suggestion
       \wedge & \vdashmemtype \idxtype~\limits'~t \ok) \\
```
Analogously with table grow.

In addition, the rules for tabletype and memtype validation (valid/types.rst) also need updating, both to handle the presence of the index type and because the size limits now depend on the index type.

bvisness replied:
Done, let me know if I missed anything.

bvisness committed:
Convert table and memory instance storage to sequences

Vectors in the spec are syntactically limited to a maximum size of 2^32.
This is incompatible with memory64, so the storage has been changed to
use sequences instead of vectors. The execution semantics are
unaffected.

bvisness commented on document/core/valid/types.rst:
```diff
@@ -501,22 +501,22 @@ Limits
 
 .. math::
    \frac{
-     C \vdashlimits \limits : 2^{32} - 1
```
I got rid of the `- 1` here. I don't think it's quite right. The limits should be about the _length_ of the sequence, not the _maximum index_. Therefore, it should be valid (semantically) to have an i32 table of length 2^32, or an i64 table of length 2^64. To bolster this point, memtype validation does not have the `- 1`.

As far as I know, this shouldn't affect any actual engine because of implementation limits (and from a pedantic point of view, it's backwards compatible).

bvisness replied:
~~To _further_ further bolster this point, the spec tests already assert that an i32 table of length 2^32 is valid: https://github.com/mozilla-spidermonkey/memory64/blob/main/test/core/table.wast#L9~~

I forgot how numbers work. Still, I think my previous points stand.

rossberg replied:
An i32 table is not allowed to have 2^32 elements, because that value is not representable as an u32, so e.g., table.size wouldn't work. Similarly with i64 tables.

In fact, 2^32 isn't even expressible, since in the binary format, the limit itself is represented using a value of the respective index type. That can't go higher than the corresponding maxuint.

bvisness replied:
Ah, I see. I suppose then that memories only get away with it because they report size in terms of pages, not bytes.

Yet another thing that might present a challenge when specifying custom page sizes, I suppose. :slightly_smiling_face:

rossberg reviewed (commented):
LGTM except for the -1 you removed, see comment.

Btw, please _never_ force-push to a branch under review. That breaks the GH review workflow, such as relative diffs.

bvisness committed:
Add back the -1 on table limits

bvisness commented:
@sbc100 Are we good to merge? I'd like to get this in so I can start working on #67.

sbc100 commented:
Sorry, I missed that you were waiting on me here.  I added you has a collaborator so you can land these kind of thing yourself going forward.
